notebooks/DenseClusImplentation.ipynb

```diff
@@ -0,0 +1,1718 @@
+    "        \"min_dist\": 0.0,\n",
+    "    },\n",
+    "}"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 14,
+   "id": "dd2b870e-fa3d-489c-afb5-32281f76f7d0",
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "categorical_umap = umap.UMAP(\n",
+    "    random_state=SEED,\n",
+    "    n_jobs=1,\n",
+    "    verbose=False,\n",
+    "    low_memory=True,\n",
+    "    **default_umap_params[\"categorical\"],\n",
+    ").fit(categorical_df)\n",
+    "\n",
+    "\n",
+    "numerical_umap = umap.UMAP(\n",
+    "    random_state=SEED,\n",
+    "    n_jobs=1,\n",
+    "    verbose=False,\n",
+    "    low_memory=True,\n",
+    "    **default_umap_params[\"numerical\"],\n",
+    ").fit(numerical_df)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 15,
+   "id": "89a8f5bb-a1f5-4597-8e7d-e1de175d6d81",
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "hdbscan_params = {\n",
+    "    \"min_cluster_size\": 100,\n",
+    "    \"min_samples\": 15,\n",
+    "    \"gen_min_span_tree\": True,\n",
+    "    \"metric\": \"euclidean\",\n",
+    "}\n",
+    "\n",
+    "hdb_numerical = hdbscan.HDBSCAN(**hdbscan_params).fit(numerical_umap.embedding_)\n",
+    "hdb_categorical = hdbscan.HDBSCAN(**hdbscan_params).fit(categorical_umap.embedding_)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 16,
+   "id": "d8eb5717-77f6-4c6c-aef8-80cd6ccbaa15",
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "DBCV numerical score: 0.279358228317536\n",
+      "DBCV categorical score: 0.4123077464785735\n"
+     ]
+    }
+   ],
+   "source": [
+    "print(f\"DBCV numerical score: {hdb_numerical.relative_validity_}\")\n",
+    "print(f\"DBCV categorical score: {hdb_categorical.relative_validity_}\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "fa2ea526-e510-4baf-a38b-ee700bfcaabc",
+   "metadata": {
+    "tags": []
+   },
+   "source": [
+    "## DenseClus on all features (numerical + categorical)\n",
+    "\n",
+    "With DenseClus this is not an issue because we use create UMAP embeddings for both categorical and numerical, combining the embedding space to output them into the densest space possible. Next HDBSCAN is run to group densities into clusters, resulting in groups of mixed-type data. \n",
+    "\n",
+    "All of this is done under the hood and just requires a `fit` call like below.\n",
+    "\n",
+    "#### There are 5 methods by which you can combine embeddings spaces (param: umap_combine_method, default=intersection)\n",
+    "\n",
+    "- 'intersection'\n",
+    "- 'union'\n",
+    "- 'contrast'\n",
+    "- 'intersection_union_mapper'\n",
+    "- 'ensemble'\n",
+    "\n",
+    "Let's implement each one of the above methods, calculate their **Density Based Clustering Validation (DBCV)** scores and visualize their respective umap embeddings. "
```

Comment:
think it would help to move dbcv description up here before showing scores

Reply:
moved the dbcv description up here.